src/languageTools/DefaultProviders.js

```diff
@@ -0,0 +1,381 @@
+            filePath: docPath,
+            cursorPos: pos
+        }).done(function (msgObj) {
+            var context = TokenUtils.getInitialContext(editor._codeMirror, pos),
+                hints = [];
+
+            self.query = context.token.string.slice(0, context.pos.ch - context.token.start);
+            if (msgObj) {
+                var res = msgObj.items,
+                    filteredHints = filterWithQueryAndMatcher(res, self.query);
+
+                StringMatch.basicMatchSort(filteredHints);
+                filteredHints.forEach(function (element) {
+                    var $fHint = $("<span>")
+                        .addClass("brackets-hints");
+
+                    if (element.stringRanges) {
+                        element.stringRanges.forEach(function (item) {
+                            if (item.matched) {
+                                $fHint.append($("<span>")
+                                    .append(_.escape(item.text))
+                                    .addClass("matched-hint"));
+                            } else {
+                                $fHint.append(_.escape(item.text));
+                            }
+                        });
+                    } else {
+                        $fHint.text(element.label);
+                    }
+
+                    $fHint.data("token", element);
+                    formatTypeDataForToken($fHint, element);
+                    hints.push($fHint);
+                });
+            }
+
+            $deferredHints.resolve({
+                "hints": hints
+            });
+        }).fail(function () {
+            $deferredHints.reject();
+        });
+
+        return $deferredHints;
+    };
+
+    CodeHintsProvider.prototype.insertHint = function ($hint) {
+        var start = {
+                line: -1,
+                ch: -1
+            },
+            end = {
+                line: -1,
+                ch: -1
+            },
+            editor = EditorManager.getActiveEditor(),
+            cursor = editor.getCursorPos(),
+            token = $hint.data("token"),
+            txt = null,
+            query = this.query;
+
+        start = {
+            line: cursor.line,
+            ch: cursor.ch - query.length
+        };
+
+        end = {
+            line: cursor.line,
+            ch: cursor.ch
+        };
+
+        txt = token.label;
+        if (token.textEdit && token.textEdit.newText) {
+            txt = token.textEdit.newText;
+            start = {
+                line: token.textEdit.range.start.line,
+                ch: token.textEdit.range.start.character
+            };
+            end = {
+                line: token.textEdit.range.end.line,
+                ch: token.textEdit.range.end.character
+            };
+        }
+
+        if (editor) {
+            editor.document.replaceRange(txt, start, end);
+        }
+        // Return false to indicate that another hinting session is not needed
```

Comment:
It can. ParameterHintsManager actually mirrors CodeHintsManager in this way. hasParameterHints basically just allows us to add any contextual inspection logic to determine if we can even have parameter hints at the location. LSP however is different, the inspection as well as inference happens in the same call, so here we just check the server abilities and return false. For non-LSP based implementation it could still return true.